test/e2e/aws_test.go

```diff
@@ -138,8 +127,297 @@ var _ = Describe("functional tests", func() {
+			})
+
+			It("should delete both clusters", func() {
+				By("Deleting the Clusters")
+				deleteCluster(ns, clName1)
+				deleteCluster(ns, clName2)
+			})
+		})
+	})
+
+	Describe("Delete machine lifecycle", func() {
+		It("It should reconcile the deleted machine", func() {
+			By("Creating a workload cluster with single control plane")
+			makeSingleControlPlaneCluster(namespace, clusterName, awsClusterName, cpAWSMachinePrefix, cpBootstrapConfigPrefix, cpMachinePrefix, instanceType, testTmpDir)
+
+			By("Creating the MachineDeployment")
+			createMachineDeployment(namespace, clusterName, machineDeploymentName, awsMachineTemplateName, mdBootstrapConfig, instanceType, initialReplicas, nil)
+
+			By("Deleting a worker node machine")
+			deleteMachine(namespace, machineDeploymentName)
+			time.Sleep(10 * time.Second)
+
+			waitForMachineDeploymentRunning(namespace, machineDeploymentName)
+
+			By("Deleting the Cluster")
+			deleteCluster(namespace, clusterName)
+		})
+	})
+
+	Describe("Workload cluster in multiple AZs", func() {
+		It("It should be creatable and deletable", func() {
+			By("Creating a workload cluster with single control plane")
+			makeSingleControlPlaneCluster(namespace, clusterName, awsClusterName, cpAWSMachinePrefix, cpBootstrapConfigPrefix, cpMachinePrefix, instanceType, testTmpDir)
+
+			ec2c := ec2.New(sess)
+			azName := getAvailabilityZone(ec2c)
+
+			By("Configuring AWS resources")
+			configureAWSresources(ec2c, clusterName, azName)
+
+			By("Creating the second Control Plane Machine in new AZ")
+			awsMachineName := cpAWSMachinePrefix + "-1"
+			bootstrapConfigName := cpBootstrapConfigPrefix + "-1"
+			machineName := cpMachinePrefix + "-1"
+			createAdditionalControlPlaneMachine(namespace, clusterName, machineName, awsMachineName, bootstrapConfigName, instanceType, azName)
+
+			By("Creating the MachineDeployment in new AZ")
+			createMachineDeployment(namespace, clusterName, machineDeploymentName, awsMachineTemplateName, mdBootstrapConfig, instanceType, initialReplicas, azName)
+
+			By("Checking if the nodes are deployed in new AZ")
+			checkInstancesInAZ(ec2c, 1+initialReplicas, azName)
+
+			By("Deleting the Cluster")
+			deleteCluster(namespace, clusterName)
+		})
+	})
 })
 
+func checkInstancesInAZ(ec2c *ec2.EC2, numOfInstances int32, azName *string) {
+	instanceInput := &ec2.DescribeInstancesInput{
+		Filters: []*ec2.Filter{
+			{
+				Name:   aws.String("availability-zone"),
+				Values: []*string{azName},
+			},
+		},
+	}
+	result, err := ec2c.DescribeInstances(instanceInput)
+	Expect(err).NotTo(HaveOccurred())
+	Expect(len(result.Reservations)).To(Equal(int(numOfInstances)))
+}
+
+func getAvailabilityZone(ec2c *ec2.EC2) *string {
+	azs, err := ec2c.DescribeAvailabilityZones(nil)
+	Expect(err).NotTo(HaveOccurred())
+	return azs.AvailabilityZones[1].ZoneName
```

Comment:
Why always the second AZ? There can be some regions with only one AZ, better check the size of the AZ list before trying to access the second one.

Reply:
yeah i did think of this before. But the tests are usually run on `us-west-2` so i just took the second one.